# **Data Collection**

## Objectives

* Fetch data from Kaggle and save it as raw data.
* Inspect the data and save it under `outputs/datasets/collection`
  
## Inputs

* Kaggle JSON file (`kaggle.json`) as the authentication token to access the kaggle dataset. 
* Raw dataset from kaggle.com for house prices in Ames, Iowa.

## Outputs

* Generate Dataset: `outputs/datasets/collection/house_prices_records.csv`

---

# Change working directory

The notebooks for this project are stored in a subfolder called `jupyter_notebooks`, therefore when running the notebook, the working directory needs to be changed to the parent folder. 
* We access the current directory with `os.getcwd()`

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/workspace/pp5-house-price-predictor/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/workspace/pp5-house-price-predictor'

---

# Fetch Data from Kaggle

Use the `%pip` magic command to install the kaggle package

In [4]:
%pip install kaggle==1.5.12

Note: you may need to restart the kernel to use updated packages.


<div style="border-left: 4px solid #ffa500; padding: 10px; background-color: #f9f9f9; color: #000">
<strong>IMPORTANT:</strong> Remember to import your <code style="color: red">kaggle.json</code> file into the root directory, as this contains the API key.
</div>

Once the `kaggle.json` file is in your root directory, run the cell below, so the token is recognized in the session

In [5]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()
! chmod 600 kaggle.json

chmod: cannot access 'kaggle.json': No such file or directory


The dataset for this project can be found at [https://www.kaggle.com/datasets/codeinstitute/housing-prices-data](https://www.kaggle.com/datasets/codeinstitute/housing-prices-data)

1. Extract the dataset path from the url. Everything after `www.kaggle.com/datasets/` and assign to a new variable
    - That being: `codeinstitute/housing-prices-data`

2. Define a new variable for the destination folder
3. Download the data from kaggle.

In [6]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "inputs/datasets/raw"
! kaggle datasets download -d {KaggleDatasetPath} -p {DestinationFolder}

Traceback (most recent call last):
  File "/workspace/.pip-modules/bin/kaggle", line 5, in <module>
    from kaggle.cli import main
  File "/workspace/.pip-modules/lib/python3.8/site-packages/kaggle/__init__.py", line 23, in <module>
    api.authenticate()
  File "/workspace/.pip-modules/lib/python3.8/site-packages/kaggle/api/kaggle_api_extended.py", line 164, in authenticate
    raise IOError('Could not find {}. Make sure it\'s located in'
OSError: Could not find kaggle.json. Make sure it's located in /workspace/pp5-house-price-predictor. Or use the environment method.


Unzip the file, remove the zip file and delete the `kaggle.json` authentication token

In [7]:
! unzip {DestinationFolder}/*.zip -d {DestinationFolder} \
    && rm {DestinationFolder}/*.zip \
    && rm kaggle.json

unzip:  cannot find or open inputs/datasets/raw/*.zip, inputs/datasets/raw/*.zip.zip or inputs/datasets/raw/*.zip.ZIP.

No zipfiles found.


---

# Load and Inspect Kaggle data

* As the file is a csv, we can use pandas to load the data into a DataFrame using the inbuilt `read_csv()` method.

In [8]:
import pandas as pd
df = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv")
df.head(10)

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   
5       796     566.0           1.0           No         732          GLQ   
6      1694       0.0           3.0           Av        1369          GLQ   
7      1107     983.0           3.0           Mn         859          ALQ   
8      1022     752.0           2.0           No           0          Unf   
9      1077       0.0           2.0           No         851          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotFrontage  \
0        150            0.0         548          RFn  ...         65.0   
1        284            NaN         460          RFn  ...         80.0   
2        434            0.0         608          RFn  ...         68.0   
3        540            NaN         642          Unf  ...         60.0   
4        490            0.0         836          RFn  ...         84.0   
5         64            NaN         480          Unf  ...         85.0   
6        317            NaN         636          RFn  ...         75.0   
7        216            NaN         484          NaN  ...          NaN   
8        952            NaN         468          Unf  ...         51.0   
9        140            NaN         205          RFn  ...         50.0   

   MasVnrArea OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  WoodDeckSF  \
0       196.0          61            5            7          856         0.0   
1         0.0           0            8            6         1262         NaN   
2       162.0          42            5            7          920         NaN   
3         0.0          35            5            7          756         NaN   
4       350.0          84            5            8         1145         NaN   
5         0.0          30            5            5          796         NaN   
6       186.0          57            5            8         1686         NaN   
7       240.0         204            6            7         1107         NaN   
8         0.0           0            5            7          952         NaN   
9         0.0           4            6            5          991         NaN   

   YearBuilt  YearRemodAdd  SalePrice  
0       2003          2003     208500  
1       1976          1976     181500  
2       2001          2002     223500  
3       1915          1970     140000  
4       2000          2000     250000  
5       1993          1995     143000  
6       2004          2005     307000  
7       1973          1973     200000  
8       1931          1950     129900  
9       1939          1950     118000  

[10 rows x 24 columns]

### Summary of Data

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       1460 non-null   int64  
 1   2ndFlrSF       1374 non-null   float64
 2   BedroomAbvGr   1361 non-null   float64
 3   BsmtExposure   1460 non-null   object 
 4   BsmtFinSF1     1460 non-null   int64  
 5   BsmtFinType1   1346 non-null   object 
 6   BsmtUnfSF      1460 non-null   int64  
 7   EnclosedPorch  136 non-null    float64
 8   GarageArea     1460 non-null   int64  
 9   GarageFinish   1298 non-null   object 
 10  GarageYrBlt    1379 non-null   float64
 11  GrLivArea      1460 non-null   int64  
 12  KitchenQual    1460 non-null   object 
 13  LotArea        1460 non-null   int64  
 14  LotFrontage    1201 non-null   float64
 15  MasVnrArea     1452 non-null   float64
 16  OpenPorchSF    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  OverallQ

### Check for Duplicate Data

* No duplicated data found. 

In [10]:
df[df.duplicated()]

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, EnclosedPorch, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, WoodDeckSF, YearBuilt, YearRemodAdd, SalePrice]
Index: []

[0 rows x 24 columns]

* 1460 records of data
* 23 Features and 1 Target variable
* 9 Columns with missing values
* Mixture of Data Types (`int64`, `float64`, `object`)
* 2 Columns with high degree of missing values
* Manual analysis of the associated text file indicates that some if not all of the object features are ordinal in nature, presenting as textual degrees of quality of a particular feature on the house. 

---